File: ios/EthereumCoin.swift
//
//  EthereumCoin.swift
//  coinswallet
//
//  Created by Tal Grynbaum on 30/06/2023.
//

import Foundation
import WalletCore

class EthereumCoin: CoinFactory.Coin {
    private var address: String = ""

    override init(mnemonic: String) {
        super.init(mnemonic: mnemonic)
        self.address = CoinType.ethereum.deriveAddress(privateKey: wallet.getKeyForCoin(coin: .ethereum))
    }

    override func getNewAddress(isTestNet:Bool) -> String {
        return self.address
    }

    override func getPrivateKey(isTestNet:Bool) -> String {
        let privateKeyBytes = wallet.getKeyForCoin(coin: .ethereum).data
        return privateKeyBytes.hexString
    }

  override func getDeriveAddresses() -> NSMutableArray {
    let result: NSMutableArray = []
    for i in 0..<50 {
          let derivePath = String(format: "m/44'/60'/0'/%@/0", String(i))
          let privateKey = wallet.getKey(coin: .ethereum, derivationPath: derivePath)
          let address = CoinType.ethereum.deriveAddress(privateKey: privateKey)
          let yourAuxDic: NSMutableDictionary = [:]
          yourAuxDic["derivePath"] = derivePath
          yourAuxDic["privateKey"] = privateKey.data.hexString
          yourAuxDic["address"] = address
          result.add(yourAuxDic)
      }
      return result;
    }
  
  override func addCustomDerivation(derivePath:String) -> NSMutableDictionary {
      let privateKey = wallet.getKey(coin: .ethereum, derivationPath: derivePath)
      let address = CoinType.ethereum.deriveAddress(privateKey: privateKey)
      let yourAuxDic: NSMutableDictionary = [:]
      yourAuxDic["derivePath"] = derivePath
      yourAuxDic["privateKey"] = privateKey.data.hexString
      yourAuxDic["address"] = address
      return yourAuxDic;
  }



    override func signTransaction(rawData: String) -> String {
        // Implement the logic to sign an Ethereum transaction using the provided raw data
        return ""
    }
}

File: android/app/src/main/java/com/coinswallet/coins/EthereumCoin.java
// EthereumCoin.java
package com.coinswallet.coins;

import com.coinswallet.CoinFactory;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.WritableMap;

import wallet.core.jni.HDWallet;
import wallet.core.jni.CoinType;
import wallet.core.jni.PrivateKey;
import java.util.Base64;
import java.nio.charset.StandardCharsets;
import android.util.Log;

public class EthereumCoin extends CoinFactory.Coin {
    private final String address;
    private final PrivateKey privateKey;

    // static {
    // Log.d("coinswallet", "Registering EthereumCoin");
    // CoinFactory.registerCoin("ethereum", EthereumCoin::new);
    // }

    public EthereumCoin(String mnemonic) {
        super(mnemonic);
        this.address = CoinType.ETHEREUM.deriveAddress(wallet.getKeyForCoin(CoinType.ETHEREUM));
        this.privateKey = wallet.getKeyForCoin(CoinType.ETHEREUM);
    }

    @Override
    public String getNewAddress(Boolean isTestNet) {
        return this.address;
    }

    @Override
    public String getPrivateKey(Boolean isTestNet) {
        return CoinFactory.Coin.toHex(this.privateKey.data());
    }

    @Override
    public ReadableArray getDeriveAddresses() {
        WritableArray results = Arguments.createArray();
       for(int i=0; i < 50; i++){
           WritableMap obj = Arguments.createMap();
           String derivePath = "m/44'/60'/0'/" + i + "/0";
           PrivateKey tempPrivateKey = wallet.getKey(CoinType.ETHEREUM,derivePath);
           String address = CoinType.ETHEREUM.deriveAddress(tempPrivateKey);
           String privateKeyString =  CoinFactory.Coin.toHex(tempPrivateKey.data());
            obj.putString("address",address);
            obj.putString("derivePath",derivePath);
            obj.putString("privateKey",privateKeyString);
            results.pushMap(obj);
       }
       return results;
    }
    @Override
    public ReadableMap addCustomDerivation(String derivePath) {
        WritableMap obj = Arguments.createMap();
        PrivateKey tempPrivateKey = wallet.getKey(CoinType.ETHEREUM,derivePath);
        String address = CoinType.ETHEREUM.deriveAddress(tempPrivateKey);
        String privateKeyString =  CoinFactory.Coin.toHex(tempPrivateKey.data());
        obj.putString("address",address);
        obj.putString("derivePath",derivePath);
        obj.putString("privateKey",privateKeyString);
        return obj;
    }

    @Override
    public String signTransaction(String rawData) {
        return null;
    }
}
